src/v/kafka/server/quota_manager.cc

```diff
@@ -124,25 +125,44 @@ std::chrono::milliseconds quota_manager::do_record_partition_mutations(
     return delay_ms;
 }
 
-// record a new observation and return <previous delay, new delay>
-throttle_delay quota_manager::record_tp_and_throttle(
-  std::optional<std::string_view> client_id,
-  uint64_t bytes,
-  clock::time_point now) {
-    auto it = maybe_add_and_retrieve_quota(client_id, now);
-
-    auto rate = it->second.tp_rate.record_and_measure(bytes, now);
-
+static std::chrono::milliseconds calculate_delay(
+  double rate, uint32_t target_rate, clock::duration window_size) {
     std::chrono::milliseconds delay_ms(0);
-    if (rate > _target_tp_rate()) {
-        auto diff = rate - _target_tp_rate();
-        double delay = (diff / _target_tp_rate())
+    if (rate > target_rate) {
+        auto diff = rate - target_rate;
+        double delay = (diff / target_rate)
                        * static_cast<double>(
                          std::chrono::duration_cast<std::chrono::milliseconds>(
-                           it->second.tp_rate.window_size())
+                           window_size)
                            .count());
         delay_ms = std::chrono::milliseconds(static_cast<uint64_t>(delay));
     }
+    return delay_ms;
+}
+
+throttle_delay quota_manager::throttle(
+  std::optional<std::string_view> client_id,
+  clock::time_point now,
+  request_type type) {
+    auto it = maybe_add_and_retrieve_quota(client_id, now);
+
+    std::chrono::milliseconds delay_ms(0);
+    double rate = 0;
+
+    if (type == request_type::produce) {
+        auto rate = it->second.tp_produce_rate.measure(now);
+        delay_ms = calculate_delay(
+          rate,
+          _target_produce_tp_rate(),
+          it->second.tp_produce_rate.window_size());
+    } else if (type == request_type::fetch && _target_fetch_tp_rate()) {
+        auto rate = it->second.tp_fetch_rate.measure(now);
+        delay_ms = calculate_delay(
+          rate,
+          *_target_fetch_tp_rate(),
+          it->second.tp_fetch_rate.window_size());
+    }
```

Comment:
I would suggest to supply `rate_tracker&` and `uint32_t target_fetch_rate` as  arguments to `quota_manager::throttle()`. Then `enum request_type` won't be needed and the code will be saved from extra branching and from the code duplication above.